agent/consul/state/catalog_events.go

```diff
@@ -0,0 +1,491 @@
+	if err != nil {
+		return nil, err
+	}
+	events = append(events, connectEvents...)
+
+	return events, nil
+}
+
+// isConnectProxyDestinationServiceChange handles the case where a Connect proxy changed
+// the service it is proxying. We need to issue a de-registration for the old
+// service on the Connect topic. We don't actually need to deregister this sidecar
+// service though as it still exists and didn't change its name.
+func isConnectProxyDestinationServiceChange(idx uint64, before, after *structs.ServiceNode) (stream.Event, bool) {
+	if before.ServiceKind != structs.ServiceKindConnectProxy ||
+		before.ServiceProxy.DestinationServiceName == after.ServiceProxy.DestinationServiceName {
+		return stream.Event{}, false
+	}
+
+	e := newServiceHealthEventDeregister(idx, before)
+	e.Topic = TopicServiceHealthConnect
+	e.Key = getPayloadCheckServiceNode(e.Payload).Service.Proxy.DestinationServiceName
+	return e, true
+}
+
+type changeType uint8
+
+const (
+	// changeIndirect indicates some other object changed which has implications
+	// for the target object.
+	changeIndirect changeType = iota
+	changeDelete
+	changeCreate
+	changeUpdate
+)
+
+func changeTypeFromChange(change memdb.Change) changeType {
+	switch {
+	case change.Deleted():
+		return changeDelete
+	case change.Created():
+		return changeCreate
+	default:
+		return changeUpdate
+	}
+}
+
+// serviceHealthToConnectEvents converts already formatted service health
+// registration events into the ones needed to publish to the Connect topic.
+// This essentially means filtering out any instances that are not Connect
+// enabled and so of no interest to those subscribers but also involves
+// switching connection details to be the proxy instead of the actual instance
+// in case of a sidecar.
+func serviceHealthToConnectEvents(tx ReadTxn, events ...stream.Event) ([]stream.Event, error) {
+	var result []stream.Event
+	for _, event := range events {
+		if event.Topic != TopicServiceHealth {
+			// Skip non-health or any events already emitted to Connect topic
+			continue
+		}
+		node := getPayloadCheckServiceNode(event.Payload)
+		if node.Service == nil {
+			continue
+		}
+
+		connectEvent := event
+		connectEvent.Topic = TopicServiceHealthConnect
+
+		switch {
+		case node.Service.Connect.Native:
+			result = append(result, connectEvent)
+
+		case node.Service.Kind == structs.ServiceKindConnectProxy:
+			connectEvent.Key = node.Service.Proxy.DestinationServiceName
+			result = append(result, connectEvent)
+
+		// TODO(streaming): add test case for Kind==ServiceKindTerminatingGateway
+		case node.Service.Kind == structs.ServiceKindTerminatingGateway:
+			// TODO(streaming): entMeta
+			_, srvs, err := gatewayServicesTxn(nil, tx, node.Service.Service, nil)
+			if err != nil {
+				return nil, err
+			}
+			for _, srv := range srvs {
+				eventCopy := connectEvent
+				eventCopy.Key = srv.Service.String()
+				result = append(result, eventCopy)
+			}
```

Comment:
I started to add this, but now I'm wondering if maybe this is the wrong way to handle it.

I imagine there are 4 cases:
1. a terminating gateway is created.
1. a service is added to a terminating gateway.
1. a service is removed from a terminating gateway.
1. a terminating gateway is removed.

Assuming that case 2 covers case 1, and case 3 covers case 4, then I think the right approach might instead be to ignore terminating gateway service changes here, and add all of the handling in the follow up to watch for changes on the `gateway-services` table.

I'm going to stash this commit somewhere for now, and revisit in the follow up.

Reply:
I'm not sure that the assumptions of case 2 covering case 1 and case 3 covering case 4 are valid.

`gateway-services` is strictly related to the config entry for the terminating gateway. If someone writes a config entry associating "my-gateway" to "amazon-rds" then an entry is created in `gateway-services`. This happens even if neither the gateway nor the database have registered service instances.

In this situation you can't send a health/connect event because "my-gateway" doesn't exist yet, and therefore there's no node/service/checks to send.

Then when an instance of "my-gateway" is registered, there won't be any updates to `gateway-services` so you would miss sending the event.

Reply:
Ahh, I missed that this table only reflected the config-entry. Thank you!

I guess both cases will require queries into the other tables to see what the current state is?

* when `gateway-services` changes, we'll need to query the catalog to see if any services have changed
* when `services` changes and kind is terminating gateway, we'll need to query `gateway-services` to see which associated services have changed

Does that sound right? I think I'll need to look into how these tables work a little more to get a better understanding. Assuming this is mostly correct I am tempted to leave all of the changes for another PR. It seems like it may require some larger changes and this PR is already pretty large.

Reply:
Yep that looks about right. What do you mean by seeing if a service changed? (last portion of each bullet above)

If `gateway-services` changes that means either:
1. A gateway is now associated with a service
2. A gateway is no longer associated with a service

For the first case we then check if that gateway exists. If it does, then that gateway's registration would be passed along to the subscriber. 

I'm not sure entirely sure how a removal would be streamed for the second case. Maybe something like "remove this service from the data you already have"